File: app/src/main/java/com/kingsoft/shiyou/omnisdk/demo/v3/java/DataMonitorApi.java
package com.kingsoft.shiyou.omnisdk.demo.v3.java;

import android.app.Activity;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;

import com.kingsoft.shiyou.omnisdk.api.entity.PayInfo;
import com.kingsoft.shiyou.omnisdk.api.entity.RoleInfo;
import com.kingsoft.shiyou.omnisdk.api.v3.OmniSDKv3;
import com.kingsoft.shiyou.omnisdk.api.v3.entity.OmniSDKRoleInfo;
import com.kingsoft.shiyou.omnisdk.api.v3.entity.events.OmniSDKCreateRoleEvent;
import com.kingsoft.shiyou.omnisdk.api.v3.entity.events.OmniSDKEnterGameEvent;
import com.kingsoft.shiyou.omnisdk.api.v3.entity.events.OmniSDKPurchaseEvent;
import com.kingsoft.shiyou.omnisdk.api.v3.entity.events.OmniSDKRevenueEvent;
import com.kingsoft.shiyou.omnisdk.api.v3.entity.events.OmniSDKRoleLevelUpEvent;
import com.kingsoft.shiyou.omnisdk.demo.v3.common.interfaces.IDataMonitorApi;
import com.kingsoft.shiyou.omnisdk.demo.v3.common.interfaces.IDataMonitorCallback;

import org.jetbrains.annotations.NotNull;

/**
 * Description: OmniSDK事件数据追踪埋点相关业务API接口代码示例Demo
 *
 * @author: LuXing created on 2021/3/24 10:30
 */
@Keep
public class DataMonitorApi implements IDataMonitorApi {

    private final String tag = "DataMonitorApi# ";
    private final Activity appActivity;
    private final IDataMonitorCallback callback;

    public DataMonitorApi(Activity activity, IDataMonitorCallback dataMonitorCallback) {
        this.appActivity = activity;
        this.callback = dataMonitorCallback;
    }

    /* ************************** OmniSDK事件数据追踪埋点功能接口示例如下 ************************** */

    public OmniSDKRoleInfo toRoleInfo(RoleInfo roleInfo) {
        return OmniSDKRoleInfo.builder()
                .userId(roleInfo.getUid())
                .roleId(roleInfo.getRoleId())
                .roleLevel(roleInfo.getRoleLevel())
                .roleName(roleInfo.getRoleName())
                .serverId(roleInfo.getServerId())
                .serverName(roleInfo.getServerName())
                .build();
    }

    /**
     * 监控追踪角色创建事件接口示例
     *
     * @param roleInfo 玩家角色信息
     * @see RoleInfo
     */
    public void onCreateRole(RoleInfo roleInfo) {
        OmniSDKCreateRoleEvent event = new OmniSDKCreateRoleEvent(toRoleInfo(roleInfo));
        OmniSDKv3.getInstance().track(event);

        callback.onSucceeded("Monitor(Tracking) `onCreateRole` Is Done");
    }

    /**
     * 监控追踪角色进入游戏事件接口示例
     *
     * @param roleInfo 玩家角色信息
     * @see RoleInfo
     */
    public void onEnterGame(RoleInfo roleInfo) {
        OmniSDKEnterGameEvent event = new OmniSDKEnterGameEvent(toRoleInfo(roleInfo));
        OmniSDKv3.getInstance().track(event);

        callback.onSucceeded("Monitor(Tracking) `onEnterGame` Is Done");
    }

    /**
     * 监控追踪角色等级升级事件接口示例
     *
     * @param roleInfo 玩家角色信息
     * @see RoleInfo
     */
    public void onRoleLevelUp(RoleInfo roleInfo) {
        OmniSDKRoleLevelUpEvent event = new OmniSDKRoleLevelUpEvent(toRoleInfo(roleInfo));
        OmniSDKv3.getInstance().track(event);

        callback.onSucceeded("Monitor(Tracking) `onRoleLevelUp` Is Done");
    }

    /**
     * 用户购买支付成功的时候调用此接口
     *
     * @param payInfo 支付相关数据
     * @see PayInfo
     */
    public void onPayFinish(PayInfo payInfo) {
        OmniSDKPurchaseEvent event = OmniSDKPurchaseEvent.builder()
                .userId(payInfo.getUid())
                .orderId(payInfo.getPlatTradeNo())
                .productId(payInfo.getProductId())
                .productDesc(payInfo.getProductDesc())
                .productUnitPrice(payInfo.getProductUnitPrice())
                .productName(payInfo.getProductName())
                .currency(payInfo.getCurrencyName())
                .gameServerId(payInfo.getServerId())
                .gameRoleVipLevel(payInfo.getRoleVipLevel())
                .gameRoleLevel(payInfo.getRoleLevel())
                .gameRoleName(payInfo.getRoleName())
                .gameRoleId(payInfo.getRoleId())
                .gameOrderId(payInfo.getGameTradeNo())
                .extJson(payInfo.getAdditionalParams())
                .purchaseQuantity(payInfo.getProductQuantity())
                .purchaseAmount(payInfo.getPayAmount())
                .build();

        OmniSDKv3.getInstance().track(event);

        callback.onSucceeded("Monitor(Tracking) `onPayFinish` Is Done");
    }

    /* ****************************************************************************************** */

    @Override
    public void onCreateRoleImpl(@NonNull RoleInfo roleInfo) {
        new Thread(() ->
                onCreateRole(roleInfo)
        ).start();
    }

    @Override
    public void onEnterGameImpl(@NonNull RoleInfo roleInfo) {
        new Thread(() ->
                onEnterGame(roleInfo)
        ).start();
    }

    @Override
    public void onRoleLevelUpImpl(@NonNull RoleInfo roleInfo) {
        new Thread(() ->
                onRoleLevelUp(roleInfo)
        ).start();
    }

    @Override
    public void onPayFinishImpl(@NonNull PayInfo payInfo) {
        new Thread(() ->
                onPayFinish(payInfo)
        ).start();
    }

    @Override
    public void dataConsumeEvent(@NotNull RoleInfo roleInfo, int consumeNum) {
        OmniSDKRevenueEvent event = OmniSDKRevenueEvent.builder()
                .roleInfo(toRoleInfo(roleInfo))
                .consumeNum(consumeNum)
                .build();
        OmniSDKv3.getInstance().track(event);
    }
}

File: app/src/main/java/com/kingsoft/shiyou/omnisdk/demo/kotlin/DataMonitorApi.kt
package com.kingsoft.shiyou.omnisdk.demo.kotlin

import android.app.Activity
import androidx.annotation.Keep
import com.kingsoft.shiyou.omnisdk.api.OmniSDK
import com.kingsoft.shiyou.omnisdk.api.entity.RoleInfo
import com.kingsoft.shiyou.omnisdk.demo.common.interfaces.IDataMonitorApi
import com.kingsoft.shiyou.omnisdk.demo.common.interfaces.IDataMonitorCallback
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Description: OMNI SDK事件数据监控追踪API接口示例Demo
 *
 * @author: LuXing created on 2021/3/24 10:00
 *
 */
@Keep
class DataMonitorApi(
    private val demoActivity: Activity,
    private val callback: IDataMonitorCallback
) :
    IDataMonitorApi {

    /* *************************** SDK 事件数据监控追踪API接口示例如下 *************************** */

    /**
     * 监控角色创建事件接口示例
     * @param roleInfo 玩家角色信息
     *
     * @see RoleInfo
     */
    private fun onCreateRole(roleInfo: RoleInfo) {
        OmniSDK.instance.onCreateRole(roleInfo)

        callback.onSucceeded("Monitor(Tracking) Event Is Done")
    }

    /**
     * 监控角色进入游戏事件接口示例
     * @param roleInfo 玩家角色信息
     *
     * @see RoleInfo
     */
    private fun onEnterGame(roleInfo: RoleInfo) {
        OmniSDK.instance.onEnterGame(roleInfo)

        callback.onSucceeded("Monitor(Tracking) Event Is Done")
    }

    /**
     * 监控角色升级事件接口示例
     * @param roleInfo 玩家角色信息
     *
     * @see RoleInfo
     */
    private fun onRoleLevelUp(roleInfo: RoleInfo) {
        OmniSDK.instance.onRoleLevelUp(roleInfo)

        callback.onSucceeded("Monitor(Tracking) Event Is Done")
    }

    /**
     * 监控游戏资源加载事件接口示例
     */
    private fun onGameLoadResource() {
        OmniSDK.instance.onGameLoadResource()

        callback.onSucceeded("Monitor(Tracking) Event Is Done")
    }

    /**
     * 监控玩家消费虚拟货币事件接口示例
     * @param roleInfo [RoleInfo]
     * @param amount 虚拟货币数量,如:10、100、100
     * @param virtualCurrencyType 虚拟货币名称，如：元宝、金币等
     * @param virtualCurrencyTotal 虚拟货币总量,如：100、1000、10000
     * @param itemName 消费项目（比如十连抽，购买体力等)
     * @param itemNum 消费项目的数量,如1、2、3
     * @param itemType 消费类型 比如道具
     *
     * @see RoleInfo
     */
    private fun onVirtualCurrencyConsume(
        roleInfo: RoleInfo,
        amount: Int,
        virtualCurrencyType: String,
        virtualCurrencyTotal: Int,
        itemName: String,
        itemNum: Int,
        itemType: String
    ) {
        OmniSDK.instance.onVirtualCurrencyConsume(
            roleInfo,
            amount,
            virtualCurrencyType,
            virtualCurrencyTotal,
            itemName,
            itemNum,
            itemType
        )

        callback.onSucceeded("Monitor(Tracking) Event Is Done")
    }

    /**
     * 监控自定义事件接口（游戏可以通过此接口进行自助数据上报及分析）
     * @param roleInfo [RoleInfo]
     * @param eventId 事件类型
     * @param eventDesc 事件描述：比如：初始化成功、登录成功等
     * @param eventVal 事件数值。XGSDK的统计平台可以对此数值进行sum／max／min／avg等操作
     * @param eventBody － 事件详细定义，JSON格式，默认为""
     *
     * @see RoleInfo
     */
    private fun onEvent(
        roleInfo: RoleInfo,
        eventId: String,
        eventDesc: String,
        eventVal: Int,
        eventBody: String
    ) {
        OmniSDK.instance.onEvent(roleInfo, eventId, eventDesc, eventVal, eventBody)

        callback.onSucceeded("Monitor(Tracking) Event Is Done")
    }

    /* ****************************************************************************************** */

    private var tag = "SDK: " + this::class.qualifiedName
    private var scope = CoroutineScope(Dispatchers.IO)

    override fun onCreateRoleImpl(roleInfo: RoleInfo) {
        scope.launch {
            onCreateRole(roleInfo)
        }
    }

    override fun onEnterGameImpl(roleInfo: RoleInfo) {
        scope.launch {
            onEnterGame(roleInfo)
        }
    }

    override fun onRoleLevelUpImpl(roleInfo: RoleInfo) {
        scope.launch {
            onRoleLevelUp(roleInfo)
        }
    }

    override fun onGameLoadResourceImpl() {
        scope.launch {
            onGameLoadResource()
        }
    }

    override fun onVirtualCurrencyConsumeImpl(
        roleInfo: RoleInfo,
        amount: Int,
        virtualCurrencyType: String,
        virtualCurrencyTotal: Int,
        itemName: String,
        itemNum: Int,
        itemType: String
    ) {
        scope.launch {
            onVirtualCurrencyConsume(
                roleInfo,
                amount,
                virtualCurrencyType,
                virtualCurrencyTotal,
                itemName,
                itemNum,
                itemType
            )
        }
    }

    override fun onEventImpl(
        roleInfo: RoleInfo,
        eventId: String,
        eventDesc: String,
        eventVal: Int,
        eventBody: String
    ) {
        scope.launch {
            onEvent(roleInfo, eventId, eventDesc, eventVal, eventBody)
        }
    }
}